From 1bc4337606180ac884e69a7e4da83cbc97422631 Mon Sep 17 00:00:00 2001
From: Robert Tiemann <r.tiemann@ta-hifi.com>
Date: Sat, 18 Apr 2020 11:52:06 +0200
Subject: [PATCH] bcm2835-i2s: Allow samples rates of up to 768000

---
 sound/soc/bcm/bcm2835-i2s.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/sound/soc/bcm/bcm2835-i2s.c b/sound/soc/bcm/bcm2835-i2s.c
index 507682d0e0bb..6f094865c6ab 100644
--- a/sound/soc/bcm/bcm2835-i2s.c
+++ b/sound/soc/bcm/bcm2835-i2s.c
@@ -774,7 +774,7 @@ static struct snd_soc_dai_driver bcm2835_i2s_dai = {
 		.channels_max = 2,
 		.rates =	SNDRV_PCM_RATE_CONTINUOUS,
 		.rate_min =	8000,
-		.rate_max =	384000,
+		.rate_max =	768000,
 		.formats =	SNDRV_PCM_FMTBIT_S16_LE
 				| SNDRV_PCM_FMTBIT_S24_LE
 				| SNDRV_PCM_FMTBIT_S32_LE
@@ -784,7 +784,7 @@ static struct snd_soc_dai_driver bcm2835_i2s_dai = {
 		.channels_max = 2,
 		.rates =	SNDRV_PCM_RATE_CONTINUOUS,
 		.rate_min =	8000,
-		.rate_max =	384000,
+		.rate_max =	768000,
 		.formats =	SNDRV_PCM_FMTBIT_S16_LE
 				| SNDRV_PCM_FMTBIT_S24_LE
 				| SNDRV_PCM_FMTBIT_S32_LE
-- 
2.30.1

