From 45be46032eea04d7c7e18ab2de41e8e82564b58a Mon Sep 17 00:00:00 2001
From: Robert Tiemann <r.tiemann@ta-hifi.com>
Date: Thu, 6 Dec 2018 14:16:11 +0100
Subject: [PATCH 2/2] Use systemd service type "simple", not "notify".

---
 bus/dbus.service.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bus/dbus.service.in b/bus/dbus.service.in
index 3713810..038e5e2 100644
--- a/bus/dbus.service.in
+++ b/bus/dbus.service.in
@@ -4,7 +4,7 @@ Documentation=man:dbus-daemon(1)
 Requires=dbus.socket
 
 [Service]
-Type=notify
+Type=simple
 NotifyAccess=main
 ExecStart=@EXPANDED_BINDIR@/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
 ExecReload=@EXPANDED_BINDIR@/dbus-send --print-reply --system --type=method_call --dest=org.freedesktop.DBus / org.freedesktop.DBus.ReloadConfig
-- 
1.9.1

