From 6b0ab55db506dd6c3147ff88ec2717a33ca6940c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Alp=20=C3=96zmert?= <info@ib-oezmert.de>
Date: Tue, 27 Nov 2018 16:42:00 +0100
Subject: [PATCH] Remove "All audio" & "All - full name".

---
 scripts/js/import.js | 21 +--------------------
 1 file changed, 1 insertion(+), 20 deletions(-)

diff --git a/scripts/js/import.js b/scripts/js/import.js
index 3ca77d57..c1e46984 100644
--- a/scripts/js/import.js
+++ b/scripts/js/import.js
@@ -120,26 +120,7 @@ function addAudio(obj) {
         }
     }
 */
-    var chain = ['Audio', 'All Audio'];
-    obj.title = title;
-    addCdsObject(obj, createContainerChain(chain));
-    
-    chain = ['Audio', 'Artists', artist, 'All Songs'];
-    addCdsObject(obj, createContainerChain(chain));
-    
-    chain = ['Audio', 'All - full name'];
-    var temp = '';
-    if (artist_full) {
-        temp = artist_full;
-    }
-    
-    if (album_full) {
-        temp = temp + ' - ' + album_full + ' - ';
-    } else {
-        temp = temp + ' - ';
-    }
-   
-    obj.title = temp + title;
+    var chain = ['Audio', 'Artists', artist, 'All Songs'];
     addCdsObject(obj, createContainerChain(chain));
     
     chain = ['Audio', 'Artists', artist, 'All - full name'];
-- 
2.17.1

