From 3e265616057ebc76d70c3ecfb6b77f4ecea6695d Mon Sep 17 00:00:00 2001
From: Robert Tiemann <r.tiemann@ta-hifi.com>
Date: Tue, 14 Jul 2020 15:00:38 +0200
Subject: [PATCH] Remove redundant YYLOC global declaration

---
 dtc-lexer.l | 1 -
 1 file changed, 1 deletion(-)

diff --git a/dtc-lexer.l b/dtc-lexer.l
index 0ee1caf..a9b58b7 100644
--- a/dtc-lexer.l
+++ b/dtc-lexer.l
@@ -38,7 +38,6 @@ LINECOMMENT	"//".*\n
 #include "srcpos.h"
 #include "dtc-parser.tab.h"
 
-YYLTYPE yylloc;
 extern bool treesource_error;
 
 /* CAUTION: this will stop working if we ever use yyless() or yyunput() */
-- 
2.27.0

