From d4f074756e7ffafbf4c385d18ccf8e34a22fe1a8 Mon Sep 17 00:00:00 2001
From: Robert Tiemann <rtie@gmx.de>
Date: Thu, 3 Dec 2020 12:07:30 +0100
Subject: [PATCH] signal: Do not use sys_siglist.

---
 common/signal.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/common/signal.c b/common/signal.c
index ccfa8e6..43648e8 100644
--- a/common/signal.c
+++ b/common/signal.c
@@ -87,11 +87,7 @@ get_signal_name( int signum )
 {
   /* Note that we can't use strsignal(), because it is not
      reentrant. */
-#if HAVE_DECL_SYS_SIGLIST && defined(NSIG)
-  return (signum >= 0 && signum < NSIG) ? sys_siglist[signum] : "?";
-#else
   return NULL;
-#endif
 }
 #endif /*!HAVE_DOSISH_SYSTEM*/
 
-- 
2.29.2

